apps/astarte_realm_management/lib/device_remover.ex

```diff
@@ -0,0 +1,207 @@
+#
+# This file is part of Astarte.
+#
+# Copyright 20230 SECO Mind Srl
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+defmodule Astarte.RealmManagement.DeviceRemover do
+  alias Astarte.RealmManagement.Queries
+
+  use Task
+  require Logger
+  alias Astarte.Core.InterfaceDescriptor
+
+  @spec start_link(list) :: {:ok, pid}
+  def start_link(args) do
+    Task.start_link(__MODULE__, :run, args)
+  end
+
+  def run(args) do
+    realm_name = Keyword.fetch(args, :realm_name)
+    device_id = Keyword.fetch(args, :device_id)
+
+    with {:ok, datastream_keys} <-
+           retrieve_individual_datastreams_keys(realm_name, device_id),
+         {:ok, _res} <- delete_individual_datastreams(realm_name, datastream_keys),
+         {:ok, property_keys} <-
+           retrieve_individual_properties_keys(realm_name, device_id),
+         {:ok, _res} <- delete_individual_properties(realm_name, property_keys),
+         {:ok, introspection} <- retrieve_device_introspection(realm_name, device_id),
+         {:ok, object_interfaces} <- filter_object_interfaces(realm_name, introspection),
+         object_tables =
+           Enum.map(object_interfaces, fn {name, major} ->
+             CQLUtils.interface_name_to_table_name(name, major)
+           end),
+         object_keys = retrieve_object_datastream_keys(realm_name, device_id, object_tables) do
+      :ok
+    end
```

Comment:
Wouldn't this just return `{:error, whatever}` if any of the stuff inside the `with` block fails and just silently fail? The error would be returned by `run`, which in turn would return it as the result of the Task, but no one is checking the result on the other side.
I think you want to convert all these functions (and queries) to bang functions.

Reply:
Moreover, if instead of having all these functions as private functions you extract them in a functional core (e.g. `DeviceRemover.Core`) you can test actually unit test them